noria-server/src/controller/migrate/mod.rs

```diff
@@ -205,47 +205,204 @@ impl<'a> Migration<'a> {
         self.mainline.graph()
     }
 
-    fn ensure_reader_for(&mut self, n: NodeIndex, name: Option<String>) {
+    fn ensure_reader_for(&mut self, n: NodeIndex, name: Option<String>, num_replicas: usize) {
         if !self.readers.contains_key(&n) {
-            // make a reader
-            let r = node::special::Reader::new(n);
-            let r = if let Some(name) = name {
-                self.mainline.ingredients[n].named_mirror(r, name)
-            } else {
-                self.mainline.ingredients[n].mirror(r)
-            };
-            let r = self.mainline.ingredients.add_node(r);
-            self.mainline.ingredients.add_edge(n, r, ());
-            self.readers.insert(n, r);
+            let mut readers = vec![];
+            for i in 0..num_replicas {
+                let replica_i = {
+                    if num_replicas == 0 {
+                        None
+                    } else {
+                        Some(i)
+                    }
+                };
+
+                // Make a reader node
+                let r = node::special::Reader::new(n, replica_i);
+                let r = if let Some(name) = name.clone() {
+                    if i == 0 {
+                        self.mainline.ingredients[n].named_mirror(r, name)
+                    } else {
+                        self.mainline.ingredients[n].named_mirror(r, format!("{}_r{}", name, i))
+                    }
+                } else {
+                    self.mainline.ingredients[n].mirror(r)
+                };
+
+                // Add it to the graph along with an edge to the node it reads for
+                let r = self.mainline.ingredients.add_node(r);
+                self.mainline.ingredients.add_edge(n, r, ());
+                self.mainline.ingredients[n].add_replica(r);
+                info!(self.log,
+                      "adding reader node";
+                      "node" => r.index(),
+                      "for_node" => n.index(),
+                      "replica_index" => replica_i
+                );
+                readers.push(r);
+            }
+
+            self.readers.insert(n, readers);
         }
     }
 
     /// Set up the given node such that its output can be efficiently queried.
     ///
     /// To query into the maintained state, use `ControllerInner::get_getter`.
     #[cfg(test)]
-    pub fn maintain_anonymous(&mut self, n: NodeIndex, key: &[usize]) -> NodeIndex {
-        self.ensure_reader_for(n, None);
-        let ri = self.readers[&n];
-
-        self.mainline.ingredients[ri]
-            .with_reader_mut(|r| r.set_key(key))
-            .unwrap();
+    pub fn maintain_anonymous(&mut self, n: NodeIndex, key: &[usize]) -> Vec<NodeIndex> {
```

Comment:
I feel like we should just kill this method and mandate that people use `maintain`...

Reply:
@fintelia @ms705 thoughts on this?

Reply:
Seems reasonable.